include/envoy/ssl/context_config.h

```diff
@@ -124,5 +136,19 @@ class ServerContextConfig : public virtual ContextConfig {
 
 using ServerContextConfigPtr = std::unique_ptr<ServerContextConfig>;
 
+/**
+ * Interface of registered factory to create TLS context config.
+ */
+class ContextConfigFactory {
+public:
+  virtual ~ContextConfigFactory() = default;
+
+  virtual ContextConfigPtr
+  createSslContextConfig(const Protobuf::Message& config,
+                         Server::Configuration::TransportSocketFactoryContext& context) PURE;
```

Comment:
It's a little strange to create a transport socket factory context when making an SSL context. Should we just define whatever context we need in this file and then have TransportSocketFactoryContext derive from that if needed?